python_modules/libraries/dagster-dbt/dagster_dbt/dbt_artifacts.py

```diff
@@ -0,0 +1,162 @@
+                Default: ["parse", "--quiet"]
+            package_data_dir (Union[str, Path]):
+                A directory that will contain a copy of the dbt project, synced at build time.
+        """
+        self.project_dir = Path(project_dir)
+        if not self.project_dir.exists():
+            raise DagsterDbtProjectNotFoundError(f"project_dir {project_dir} does not exist.")
+        self._target_folder = Path(target_folder)
+        self._prepare_command = prepare_command
+        # this is ok if it doesn't exist, will get created by `prepare`
+        self._package_data_dir = Path(package_data_dir) if package_data_dir else None
+
+        if self._should_prepare_at_runtime():
+            self._manifest_path = self._prepare_manifest()
+        else:
+            self._manifest_path = self._base_dir.joinpath(self._target_folder, "manifest.json")
+
+    @public
+    @property
+    def manifest_path(self) -> Path:
+        """The path to the manifest.json, compiling the manifest first if in development or
+        ensuring it already exists if not.
+        """
+        if not self._manifest_path.exists():
+            if self._should_prepare_at_runtime():
+                raise DagsterDbtManifestNotPreparedError(
+                    f"Unexpected state, {self._manifest_path} should have been created when this "
+                    "object was initialized."
+                )
+            else:
+                raise DagsterDbtManifestNotPreparedError(
+                    f"Did not find prepared manifest.json at expected path {self._manifest_path}.\n"
+                    "If this is an environment that is expected to prepare the manifest at run time, "
+                    "set the environment variable DAGSTER_DBT_PARSE_PROJECT_ON_LOAD."
+                )
+
+        return self._manifest_path
+
+    @public
+    def prepare(self, *, quiet=False) -> None:
+        """A method that can be called as part of the deployment process to handle
+        preparing the manifest and if package_data_dir is set, handle copying
+        the dbt project to that directory.
+
+        Args:
+            quiet (bool):
+                Disable logging
+                Default: False
+        """
+        logger = _NULL_LOGGER if quiet else logging.getLogger("dagster-dbt-artifacts")
+        self._prepare_manifest(logger)
+        self._handle_package_data(logger)
+        logger.info("Preparation complete.")
+
+    @public
+    def get_cli_resource(self, **kwargs) -> DbtCliResource:
+        """Construct a DbtCliResource that targets the correct project directory, passing any
+        kwargs to DbtCliResource.
+        If package_data_dir is set, resolves to that outside of the development context and
+        uses project_dir when in a development context.
+        """
+        if self._should_prepare_at_runtime():
+            return DbtCliResource(project_dir=os.fspath(self.project_dir))
```

Comment:
should this get **kwargs passed in as well?

Reply:
Related to my larger comment at the top but I'm also finding this confusing as to why _should_prepare_at_runtime is related to what path this should be pointing at. Functionally, I get that the method checks if you're currently running from a `dagster dev` invocation, but that seems like a separate concern from if `DAGSTER_DBT_PARSE_PROJECT_ON_LOAD` is set or not. I get that it'd be fairly unlikely to set that outside of a development context, but if we want to make that assertion we should just call this method "_is_local_environment" or something.

That'd make other code paths clearer as well (took me a bit to wrap my head around _should_prepare_at_runtime = this code will only run when in a dev environment)